wavefront/resource_service_account.go

```diff
@@ -0,0 +1,169 @@
+package wavefront
+
+import (
+	"fmt"
+
+	"github.com/WavefrontHQ/go-wavefront-management-api"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceServiceAccount() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceServiceAccountCreate,
+		Read:   resourceServiceAccountRead,
+		Update: resourceServiceAccountUpdate,
+		Delete: resourceServiceAccountDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"identifier": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"active": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"permissions": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
+				Computed: true,
+				Set:      schema.HashString,
+			},
+			"user_groups": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
+				Computed: true,
+				Set:      schema.HashString,
+			},
+		},
+	}
+}
+
+func resourceServiceAccountCreate(
+	d *schema.ResourceData, meta interface{}) error {
+	serviceAccounts := meta.(*wavefrontClient).client.ServiceAccounts()
+	tokens := meta.(*wavefrontClient).client.Tokens()
+
+	serviceAccount, err := serviceAccounts.Create(
+		&wavefront.ServiceAccountOptions{
+			ID:          d.Get("identifier").(string),
+			Active:      d.Get("active").(bool),
+			Description: d.Get("description").(string),
+			Permissions: getStringSlice(d, "permissions"),
+			UserGroups:  getStringSlice(d, "user_groups"),
+		})
+	if err != nil {
+		return fmt.Errorf(
+			"failed to create new Wavefront Service Account, %s",
+			err)
+	}
+	_, err = tokens.Create(serviceAccount.ID, &wavefront.TokenOptions{Name: "main"})
+	if err != nil {
+		return fmt.Errorf(
+			"failed to create token for new Wavefront ServiceAccount, %s",
+			err)
+	}
+	d.SetId(serviceAccount.ID)
+	err = setStringSlice(d, "permissions", serviceAccount.Permissions)
+	if err != nil {
+		return err
+	}
+	return setStringSlice(d, "user_groups", serviceAccount.UserGroupIds())
```

Comment:
A better best practice is for Create and Update to return a call to the Read function to ensure a common state. Here's an [example](https://github.com/vmware/terraform-provider-vra/blob/master/vra/resource_project.go#L158) from the vra provider.